modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPartitionTimestampCursor.java

```diff
@@ -0,0 +1,185 @@
+        this.storage = storage;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return storage.busy(() -> {
+            storage.throwExceptionIfStorageNotInRunnableState();
+
+            if (nextRead != null) {
+                return true;
+            }
+
+            if (iterationExhausted) {
+                return false;
+            }
+
+            createVersionChainCursorIfMissing();
+
+            currentChain = null;
+
+            while (true) {
+                if (!versionChainCursor.hasNext()) {
+                    iterationExhausted = true;
+
+                    return false;
+                }
+
+                VersionChain chain = versionChainCursor.next();
+
+                ReadResult result = readResultByRowId.remove(chain.rowId());
+
+                if (result == null) {
+                    // TODO: IGNITE-18717 Add lock by rowId
+                    chain = storage.readVersionChain(chain.rowId());
+
+                    if (chain == null) {
+                        continue;
+                    }
+
+                    result = findRowVersion(chain);
+                }
+
+                if (result.isEmpty() && !result.isWriteIntent()) {
+                    continue;
+                }
+
+                nextRead = result;
+                currentChain = chain;
+
+                return true;
+            }
+        });
+    }
+
+    @Override
+    public final ReadResult next() {
+        return storage.busy(() -> {
+            storage.throwExceptionIfStorageNotInRunnableState();
+
+            if (!hasNext()) {
+                throw new NoSuchElementException("The cursor is exhausted: " + storage.createStorageInfo());
+            }
+
+            assert nextRead != null;
+
+            ReadResult res = nextRead;
+
+            nextRead = null;
+
+            return res;
+        });
+    }
+
+    @Override
+    public void close() {
+        if (versionChainCursor != null) {
+            versionChainCursor.close();
+        }
+    }
+
+    @Override
+    public @Nullable TableRow committed(HybridTimestamp timestamp) {
+        return storage.busy(() -> {
+            storage.throwExceptionIfStorageNotInRunnableState();
+
+            if (currentChain == null) {
+                throw new IllegalStateException("Version chain missing: " + storage.createStorageInfo());
+            }
+
+            // TODO: IGNITE-18717 Add lock by rowId
+            VersionChain chain = storage.readVersionChain(currentChain.rowId());
+
+            if (chain == null) {
+                return null;
+            }
+
+            ReadResult result = storage.findRowVersionByTimestamp(chain, timestamp);
+
+            if (result.isEmpty()) {
+                return null;
+            }
+
+            // We don't check if row conforms the key filter here, because we've already checked it.
+            return result.tableRow();
+        });
+    }
+
+    abstract ReadResult findRowVersion(VersionChain versionChain);
```

Comment:
I think it would be nice to have a javadoc here (explaining what  this method does)